Modules/_functoolsmodule.c

```diff
@@ -40,9 +67,15 @@ partial_new(PyTypeObject *type, PyObject *args, PyObject *kw)
         return NULL;
     }
 
+    _functools_state *state = get_functools_state_by_type(type);
+    if (state == NULL) {
+        return NULL;
+    }
+
     pargs = pkw = NULL;
     func = PyTuple_GET_ITEM(args, 0);
-    if (Py_IS_TYPE(func, &partial_type) && type == &partial_type) {
+    if (Py_IS_TYPE(func, state->partial_type)
+        && type == state->partial_type) {
```

Comment:
This can be simplified. We don't actually need to check the exact type, we just need a particular memory layout.
```suggestion
    if (Py_TYPE(func)->tp_call == partial_call) {
        // The type of "func" might not be exactly the same type object
        // as "type", but if it is called using partial_call, it must have the
        // same memory layout (fn, args and kw members).
        // We can use its underlying function directly and merge the arguments.
```
This will need a forward declaration of `partial_call`, but it would mean we don't need `get_functools_state_by_type` in this function and we don't need to store `partial_type` in the module state.

cc @vstinner

Reply:
> We don't actually need to check the exact type, we just need a particular memory layout.
I am not sure about it. If it's no backward compatibility, I can update it.

Reply:
How about using `Py_TYPE(func)->tp_new == partial_new` in here?

Reply:
I picked `tp_call` because we care about how the object is called, not about how it's created.
By nature, `partial_call` uses all the C API needed here. It's also smaller than `partial_new`; it uses *only* the API needed here.

Reply:
OK, Thanks. updated.